### ISSUE TITLE
StackOverflow on running compute shader

### ISSUE THREAD

**Issue by KallDrexx** [Sergio0694/ComputeSharp#119]
I'm playing around with benchmarking architectural changes for my C# particle engine, and for kicks I decided to see what timing a compute shader implementation would look like.  So I created the following code as a best guess from the current docs:

[ComputeShaderRepro.zip](https://github.com/Sergio0694/ComputeSharp/files/6306738/ComputeShaderRepro.zip)

However, when this runs I get:

```
Stack overflow.
   at TerraFX.Interop.IDxcCompiler.Compile(TerraFX.Interop.IDxcBlob*, UInt16*, UInt16*, UInt16*, UInt16**, UInt32, TerraFX.Interop.DxcDefine*, UInt32, TerraFX.Interop.IDxcIncludeHandler*, TerraFX.Interop.IDxcOperationResult**)
   at ComputeSharp.Shaders.Translation.ShaderCompiler.CompileShader(System.ReadOnlySpan`1<Char>)
   at ComputeSharp.Shaders.ShaderRunner`1[[ParticleBenchmark.Shader, ParticleBenchmark, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].LoadShader(Int32, Int32, Int32, ParticleBenchmark.Shader ByRef, ComputeSharp.Shaders.Translation.Models.CachedShader`1<ParticleBenchmark.Shader> ByRef)
   at ComputeSharp.Shaders.ShaderRunner`1[[ParticleBenchmark.Shader, ParticleBenchmark, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].Run(ComputeSharp.GraphicsDevice, Int32, Int32, Int32, Int32, Int32, Int32, ParticleBenchmark.Shader ByRef)
   at ComputeSharp.Shaders.ShaderRunner`1[[ParticleBenchmark.Shader, ParticleBenchmark, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]].Run(ComputeSharp.GraphicsDevice, Int32, Int32, Int32, ParticleBenchmark.Shader ByRef)
   at ComputeSharp.GraphicsDeviceExtensions.For[[ParticleBenchmark.Shader, ParticleBenchmark, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null]](ComputeSharp.GraphicsDevice, Int32, ParticleBenchmark.Shader ByRef)
   at ParticleBenchmark.ComputeShaderParticles+Emitter.Update(Single)
   at ParticleBenchmark.Program.Main(System.String[])

```

I'm not really sure what to look at here to address this.  This is using the `2.0.0-alpha.2` nuget package, which seems recent.  Any ideas on what I"m doing wrong?

**Comment by KallDrexx** [comment]
FWIW   Itried removing the passed in values, I tried using a `struct Particle` so I wasn't passing in many different buffers, but I couldn't make any progress.

**Comment by Sergio0694** [comment]
Hi @KallDrexx, it's super cool that you tried to use ComputeSharp in your own particle engine! 😄

The reason for the crash in this case is that you're using a buffer of type `byte`, which is not supported (see the full list of supported types in the readme). In this case, a simple fix is to make that buffer an `int` buffer, and to also change that `(byte)` cast in the shader body to be `(int)`, possibly by adding an `Hlsl.Clamp(..., 0, 255)` call if you need to ensure the range there. I did that fix locally and the shader runs fine after that 🙂

Note: you can get rid of that huge constructor in your shader type by just making the shader type `partial`, and adding `[AutoConstructor]` on it, then ComputeSharp will create a constructor for you to assign all those declared fields 👍

Will keep this issue open until I've pushed a fix to properly handle invalid types being used, instead of just having a crash like this.
Let me know if it works for you with this change, and also how is the performance in your particle engine!

**Comment by KallDrexx** [comment]
> The reason for the crash in this case is that you're using a buffer of type `byte`, which is not supported (see the full list of supported types in the readme). In this case, a simple fix is to make that buffer an `int` buffer, and to also change that `(byte)` cast in the shader body to be `(int)`, possibly by adding an `Hlsl.Clamp(..., 0, 255)` call if you need to ensure the range there. I did that fix locally and the shader runs fine after that 🙂

Aha that worked!  I only used byte for those anyway for memory considerations, not because I necessarily need the values clamped (its rare it'll go over 20, let alone 255).

> Note: you can get rid of that huge constructor in your shader type by just making the shader type partial, and adding [AutoConstructor] on it, then ComputeSharp will create a constructor for you to assign all those declared fields 👍

I tried that but it didn't generate any constructors for me.  I wasn't sure at if I had to call the constructor first to generate it or what (and how it matched things up) so I just removed it and went with a manual constructor.

Thanks for the swift reply :)

**Comment by Sergio0694** [comment]
> "I tried that but it didn't generate any constructors for me. I wasn't sure at if I had to call the constructor first to generate it or what (and how it matched things up) so I just removed it and went with a manual constructor."

If you see the red squiggles, that's just because the tooling isn't still perfect in Visual Studio yet (still in preview). If you try using the constructor anyway and build and run, it should actually work fine even if IntelliSense thinks the constructor doesn't exist 🙂

**Comment by Sergio0694** [comment]
Sorry for the delay! This is now fixed as of e54ba8f889b1814e549359581a346f08e8159252. Thank you! 🙂
